File: src/libqoperatingsystemversion.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");
const qoperatingsystemversion_enums = enums;
const std = @import("std");

/// https://doc.qt.io/qt-6/qoperatingsystemversionbase.html
pub const qoperatingsystemversionbase = struct {
    /// New constructs a new QOperatingSystemVersionBase object.
    ///
    /// ``` osType: qoperatingsystemversion_enums.OSType, vmajor: i32 ```
    pub fn New(osType: i64, vmajor: i32) QtC.QOperatingSystemVersionBase {
        return qtc.QOperatingSystemVersionBase_new(@intCast(osType), @intCast(vmajor));
    }

    /// New2 constructs a new QOperatingSystemVersionBase object.
    ///
    /// ``` param1: QtC.QOperatingSystemVersionBase ```
    pub fn New2(param1: ?*anyopaque) QtC.QOperatingSystemVersionBase {
        return qtc.QOperatingSystemVersionBase_new2(@ptrCast(param1));
    }

    /// New3 constructs a new QOperatingSystemVersionBase object.
    ///
    /// ``` osType: qoperatingsystemversion_enums.OSType, vmajor: i32, vminor: i32 ```
    pub fn New3(osType: i64, vmajor: i32, vminor: i32) QtC.QOperatingSystemVersionBase {
        return qtc.QOperatingSystemVersionBase_new3(@intCast(osType), @intCast(vmajor), @intCast(vminor));
    }

    /// New4 constructs a new QOperatingSystemVersionBase object.
    ///
    /// ``` osType: qoperatingsystemversion_enums.OSType, vmajor: i32, vminor: i32, vmicro: i32 ```
    pub fn New4(osType: i64, vmajor: i32, vminor: i32, vmicro: i32) QtC.QOperatingSystemVersionBase {
        return qtc.QOperatingSystemVersionBase_new4(@intCast(osType), @intCast(vmajor), @intCast(vminor), @intCast(vmicro));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qoperatingsystemversionbase.html#current)
    ///
    ///
    pub fn Current() QtC.QOperatingSystemVersionBase {
        return qtc.QOperatingSystemVersionBase_Current();
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qoperatingsystemversionbase.html#name)
    ///
    /// ``` osversion: QtC.QOperatingSystemVersionBase, allocator: std.mem.Allocator ```
    pub fn Name(osversion: QtC.QOperatingSystemVersionBase, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QOperatingSystemVersionBase_Name(@ptrCast(osversion));
        defer qtc.libqt_string_free(@constCast(&_str));
        const _ret = allocator.alloc(u8, _str.len) catch @panic("qoperatingsystemversionbase.Name: Memory allocation failed");
        for (0.._str.len) |_i| {
            _ret[_i] = _str.data[_i];
        }
        return _ret;
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qoperatingsystemversionbase.html#currentType)
    ///
    ///
    pub fn CurrentType() i64 {
        return qtc.QOperatingSystemVersionBase_CurrentType();
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qoperatingsystemversionbase.html#version)
    ///
    /// ``` self: QtC.QOperatingSystemVersionBase ```
    pub fn Version(self: ?*anyopaque) QtC.QVersionNumber {
        return qtc.QOperatingSystemVersionBase_Version(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qoperatingsystemversionbase.html#majorVersion)
    ///
    /// ``` self: QtC.QOperatingSystemVersionBase ```
    pub fn MajorVersion(self: ?*anyopaque) i32 {
        return qtc.QOperatingSystemVersionBase_MajorVersion(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qoperatingsystemversionbase.html#minorVersion)
    ///
    /// ``` self: QtC.QOperatingSystemVersionBase ```
    pub fn MinorVersion(self: ?*anyopaque) i32 {
        return qtc.QOperatingSystemVersionBase_MinorVersion(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qoperatingsystemversionbase.html#microVersion)
    ///
    /// ``` self: QtC.QOperatingSystemVersionBase ```
    pub fn MicroVersion(self: ?*anyopaque) i32 {
        return qtc.QOperatingSystemVersionBase_MicroVersion(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qoperatingsystemversionbase.html#segmentCount)
    ///
    /// ``` self: QtC.QOperatingSystemVersionBase ```
    pub fn SegmentCount(self: ?*anyopaque) i32 {
        return qtc.QOperatingSystemVersionBase_SegmentCount(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qoperatingsystemversionbase.html#type)
    ///
    /// ``` self: QtC.QOperatingSystemVersionBase ```
    pub fn Type(self: ?*anyopaque) i64 {
        return qtc.QOperatingSystemVersionBase_Type(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qoperatingsystemversionbase.html#name)
    ///
    /// ``` self: QtC.QOperatingSystemVersionBase, allocator: std.mem.Allocator ```
    pub fn Name2(self: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QOperatingSystemVersionBase_Name2(@ptrCast(self));
        defer qtc.libqt_string_free(@constCast(&_str));
        const _ret = allocator.alloc(u8, _str.len) catch @panic("qoperatingsystemversionbase.Name2: Memory allocation failed");
        for (0.._str.len) |_i| {
            _ret[_i] = _str.data[_i];
        }
        return _ret;
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qoperatingsystemversionbase.html#dtor.QOperatingSystemVersionBase)
    ///
    /// Delete this object from C++ memory.
    ///
    /// ``` self: QtC.QOperatingSystemVersionBase ```
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QOperatingSystemVersionBase_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/qoperatingsystemversion.html
pub const qoperatingsystemversion = struct {
    /// New constructs a new QOperatingSystemVersion object.
    ///
    /// ``` osversion: QtC.QOperatingSystemVersionBase ```
    pub fn New(osversion: ?*anyopaque) QtC.QOperatingSystemVersion {
        return qtc.QOperatingSystemVersion_new(@ptrCast(osversion));
    }

    /// New2 constructs a new QOperatingSystemVersion object.
    ///
    /// ``` osType: qoperatingsystemversion_enums.OSType, vmajor: i32 ```
    pub fn New2(osType: i64, vmajor: i32) QtC.QOperatingSystemVersion {
        return qtc.QOperatingSystemVersion_new2(@intCast(osType), @intCast(vmajor));
    }

    /// New3 constructs a new QOperatingSystemVersion object.
    ///
    /// ``` param1: QtC.QOperatingSystemVersion ```
    pub fn New3(param1: ?*anyopaque) QtC.QOperatingSystemVersion {
        return qtc.QOperatingSystemVersion_new3(@ptrCast(param1));
    }

    /// New4 constructs a new QOperatingSystemVersion object.
    ///
    /// ``` osType: qoperatingsystemversion_enums.OSType, vmajor: i32, vminor: i32 ```
    pub fn New4(osType: i64, vmajor: i32, vminor: i32) QtC.QOperatingSystemVersion {
        return qtc.QOperatingSystemVersion_new4(@intCast(osType), @intCast(vmajor), @intCast(vminor));
    }

    /// New5 constructs a new QOperatingSystemVersion object.
    ///
    /// ``` osType: qoperatingsystemversion_enums.OSType, vmajor: i32, vminor: i32, vmicro: i32 ```
    pub fn New5(osType: i64, vmajor: i32, vminor: i32, vmicro: i32) QtC.QOperatingSystemVersion {
        return qtc.QOperatingSystemVersion_new5(@intCast(osType), @intCast(vmajor), @intCast(vminor), @intCast(vmicro));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qoperatingsystemversion.html#current)
    ///
    ///
    pub fn Current() QtC.QOperatingSystemVersion {
        return qtc.QOperatingSystemVersion_Current();
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qoperatingsystemversion.html#currentType)
    ///
    ///
    pub fn CurrentType() i64 {
        return qtc.QOperatingSystemVersion_CurrentType();
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qoperatingsystemversion.html#version)
    ///
    /// ``` self: QtC.QOperatingSystemVersion ```
    pub fn Version(self: ?*anyopaque) QtC.QVersionNumber {
        return qtc.QOperatingSystemVersion_Version(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qoperatingsystemversion.html#majorVersion)
    ///
    /// ``` self: QtC.QOperatingSystemVersion ```
    pub fn MajorVersion(self: ?*anyopaque) i32 {
        return qtc.QOperatingSystemVersion_MajorVersion(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qoperatingsystemversion.html#minorVersion)
    ///
    /// ``` self: QtC.QOperatingSystemVersion ```
    pub fn MinorVersion(self: ?*anyopaque) i32 {
        return qtc.QOperatingSystemVersion_MinorVersion(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qoperatingsystemversion.html#microVersion)
    ///
    /// ``` self: QtC.QOperatingSystemVersion ```
    pub fn MicroVersion(self: ?*anyopaque) i32 {
        return qtc.QOperatingSystemVersion_MicroVersion(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qoperatingsystemversion.html#segmentCount)
    ///
    /// ``` self: QtC.QOperatingSystemVersion ```
    pub fn SegmentCount(self: ?*anyopaque) i32 {
        return qtc.QOperatingSystemVersion_SegmentCount(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qoperatingsystemversion.html#type)
    ///
    /// ``` self: QtC.QOperatingSystemVersion ```
    pub fn Type(self: ?*anyopaque) i64 {
        return qtc.QOperatingSystemVersion_Type(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qoperatingsystemversion.html#name)
    ///
    /// ``` self: QtC.QOperatingSystemVersion, allocator: std.mem.Allocator ```
    pub fn Name(self: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QOperatingSystemVersion_Name(@ptrCast(self));
        defer qtc.libqt_string_free(@constCast(&_str));
        const _ret = allocator.alloc(u8, _str.len) catch @panic("qoperatingsystemversion.Name: Memory allocation failed");
        for (0.._str.len) |_i| {
            _ret[_i] = _str.data[_i];
        }
        return _ret;
    }

    /// Inherited from QOperatingSystemVersionBase
    ///
    /// [Qt documentation](https://doc.qt.io/qt-6/qoperatingsystemversionbase.html#name)
    ///
    /// ``` self: QtC.QOperatingSystemVersion, allocator: std.mem.Allocator ```
    pub fn Name2(self: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QOperatingSystemVersionBase_Name2(@ptrCast(self));
        defer qtc.libqt_string_free(@constCast(&_str));
        const _ret = allocator.alloc(u8, _str.len) catch @panic("qoperatingsystemversion.Name2: Memory allocation failed");
        for (0.._str.len) |_i| {
            _ret[_i] = _str.data[_i];
        }
        return _ret;
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qoperatingsystemversion.html#dtor.QOperatingSystemVersion)
    ///
    /// Delete this object from C++ memory.
    ///
    /// ``` self: QtC.QOperatingSystemVersion ```
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QOperatingSystemVersion_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/qoperatingsystemversion.html#types
pub const enums = struct {
    pub const QOperatingSystemVersionBase__OSType = enum {
        pub const Unknown: i32 = 0;
        pub const Windows: i32 = 1;
        pub const MacOS: i32 = 2;
        pub const IOS: i32 = 3;
        pub const TvOS: i32 = 4;
        pub const WatchOS: i32 = 5;
        pub const Android: i32 = 6;
    };

    pub const QOperatingSystemVersion__OSType = enum {
        pub const Unknown: i32 = 0;
        pub const Windows: i32 = 1;
        pub const MacOS: i32 = 2;
        pub const IOS: i32 = 3;
        pub const TvOS: i32 = 4;
        pub const WatchOS: i32 = 5;
        pub const Android: i32 = 6;
    };
};

File: src/libqoperatingsystemversion.cpp
#include <QOperatingSystemVersion>
#include <QOperatingSystemVersionBase>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <QVersionNumber>
#include <qoperatingsystemversion.h>
#include "libqoperatingsystemversion.h"
#include "libqoperatingsystemversion.hxx"

QOperatingSystemVersionBase* QOperatingSystemVersionBase_new(int osType, int vmajor) {
    return new QOperatingSystemVersionBase(static_cast<QOperatingSystemVersionBase::OSType>(osType), static_cast<int>(vmajor));
}

QOperatingSystemVersionBase* QOperatingSystemVersionBase_new2(const QOperatingSystemVersionBase* param1) {
    return new QOperatingSystemVersionBase(*param1);
}

QOperatingSystemVersionBase* QOperatingSystemVersionBase_new3(int osType, int vmajor, int vminor) {
    return new QOperatingSystemVersionBase(static_cast<QOperatingSystemVersionBase::OSType>(osType), static_cast<int>(vmajor), static_cast<int>(vminor));
}

QOperatingSystemVersionBase* QOperatingSystemVersionBase_new4(int osType, int vmajor, int vminor, int vmicro) {
    return new QOperatingSystemVersionBase(static_cast<QOperatingSystemVersionBase::OSType>(osType), static_cast<int>(vmajor), static_cast<int>(vminor), static_cast<int>(vmicro));
}

QOperatingSystemVersionBase* QOperatingSystemVersionBase_Current() {
    return new QOperatingSystemVersionBase(QOperatingSystemVersionBase::current());
}

libqt_string QOperatingSystemVersionBase_Name(QOperatingSystemVersionBase* osversion) {
    QString _ret = QOperatingSystemVersionBase::name(*osversion);
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<char*>(malloc((_str.len + 1) * sizeof(char)));
    memcpy(_str.data, _b.data(), _str.len);
    _str.data[_str.len] = '\0';
    return _str;
}

int QOperatingSystemVersionBase_CurrentType() {
    return static_cast<int>(QOperatingSystemVersionBase::currentType());
}

QVersionNumber* QOperatingSystemVersionBase_Version(const QOperatingSystemVersionBase* self) {
    return new QVersionNumber(self->version());
}

int QOperatingSystemVersionBase_MajorVersion(const QOperatingSystemVersionBase* self) {
    return self->majorVersion();
}

int QOperatingSystemVersionBase_MinorVersion(const QOperatingSystemVersionBase* self) {
    return self->minorVersion();
}

int QOperatingSystemVersionBase_MicroVersion(const QOperatingSystemVersionBase* self) {
    return self->microVersion();
}

int QOperatingSystemVersionBase_SegmentCount(const QOperatingSystemVersionBase* self) {
    return self->segmentCount();
}

int QOperatingSystemVersionBase_Type(const QOperatingSystemVersionBase* self) {
    return static_cast<int>(self->type());
}

libqt_string QOperatingSystemVersionBase_Name2(const QOperatingSystemVersionBase* self) {
    QString _ret = self->name();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<char*>(malloc((_str.len + 1) * sizeof(char)));
    memcpy(_str.data, _b.data(), _str.len);
    _str.data[_str.len] = '\0';
    return _str;
}

void QOperatingSystemVersionBase_Delete(QOperatingSystemVersionBase* self) {
    delete self;
}

QOperatingSystemVersion* QOperatingSystemVersion_new(const QOperatingSystemVersionBase* osversion) {
    return new QOperatingSystemVersion(*osversion);
}

QOperatingSystemVersion* QOperatingSystemVersion_new2(int osType, int vmajor) {
    return new QOperatingSystemVersion(static_cast<QOperatingSystemVersion::OSType>(osType), static_cast<int>(vmajor));
}

QOperatingSystemVersion* QOperatingSystemVersion_new3(const QOperatingSystemVersion* param1) {
    return new QOperatingSystemVersion(*param1);
}

QOperatingSystemVersion* QOperatingSystemVersion_new4(int osType, int vmajor, int vminor) {
    return new QOperatingSystemVersion(static_cast<QOperatingSystemVersion::OSType>(osType), static_cast<int>(vmajor), static_cast<int>(vminor));
}

QOperatingSystemVersion* QOperatingSystemVersion_new5(int osType, int vmajor, int vminor, int vmicro) {
    return new QOperatingSystemVersion(static_cast<QOperatingSystemVersion::OSType>(osType), static_cast<int>(vmajor), static_cast<int>(vminor), static_cast<int>(vmicro));
}

QOperatingSystemVersion* QOperatingSystemVersion_Current() {
    return new QOperatingSystemVersion(QOperatingSystemVersion::current());
}

int QOperatingSystemVersion_CurrentType() {
    return static_cast<int>(QOperatingSystemVersion::currentType());
}

QVersionNumber* QOperatingSystemVersion_Version(const QOperatingSystemVersion* self) {
    return new QVersionNumber(self->version());
}

int QOperatingSystemVersion_MajorVersion(const QOperatingSystemVersion* self) {
    return self->majorVersion();
}

int QOperatingSystemVersion_MinorVersion(const QOperatingSystemVersion* self) {
    return self->minorVersion();
}

int QOperatingSystemVersion_MicroVersion(const QOperatingSystemVersion* self) {
    return self->microVersion();
}

int QOperatingSystemVersion_SegmentCount(const QOperatingSystemVersion* self) {
    return self->segmentCount();
}

int QOperatingSystemVersion_Type(const QOperatingSystemVersion* self) {
    return static_cast<int>(self->type());
}

libqt_string QOperatingSystemVersion_Name(const QOperatingSystemVersion* self) {
    QString _ret = self->name();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<char*>(malloc((_str.len + 1) * sizeof(char)));
    memcpy(_str.data, _b.data(), _str.len);
    _str.data[_str.len] = '\0';
    return _str;
}

void QOperatingSystemVersion_Delete(QOperatingSystemVersion* self) {
    delete self;
}
